File: src/hardware_splicer/engineering_review_api.py
"""Product routes for optional external engineering review adapters."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from .integrations.engineering_review import (
    engineering_review_status,
    run_engineering_review,
)


class EngineeringReviewBuildRequest(BaseModel):
    build_dir: str


class EngineeringReviewRunRequest(EngineeringReviewBuildRequest):
    timeout_s: float = Field(default=180.0, ge=1.0, le=1800.0)
    force: bool = False


def _validation_error(exc: ValueError) -> HTTPException:
    return HTTPException(
        status_code=422,
        detail={
            "error": {
                "code": "engineering_review_validation_error",
                "message": str(exc),
            }
        },
    )


def create_engineering_review_router() -> APIRouter:
    router = APIRouter(
        prefix="/v1/build-files/engineering-review",
        tags=["engineering-review"],
    )

    @router.post("/status")
    def status(request: EngineeringReviewBuildRequest) -> dict[str, Any]:
        try:
            return engineering_review_status(request.build_dir)
        except ValueError as exc:
            raise _validation_error(exc) from exc

    @router.post("/run")
    def run(request: EngineeringReviewRunRequest) -> dict[str, Any]:
        try:
            return run_engineering_review(
                request.build_dir,
                timeout_s=request.timeout_s,
                force=request.force,
            )
        except ValueError as exc:
            raise _validation_error(exc) from exc

    return router

File: apps/splice-ui/src/engineeringReviewApi.js
const API_BASE =
  import.meta.env.VITE_API_BASE !== undefined
    ? import.meta.env.VITE_API_BASE
    : import.meta.env.DEV
      ? "/api"
      : "";

async function parseJson(res) {
  const body = await res.json().catch(() => ({}));
  if (!res.ok) {
    const detail = body?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : detail?.error?.message || detail?.message || res.statusText;
    throw new Error(message || `Request failed (${res.status})`);
  }
  return body;
}

export async function fetchEngineeringReviewStatus(buildDir) {
  const res = await fetch(`${API_BASE}/v1/build-files/engineering-review/status`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ build_dir: buildDir }),
  });
  return parseJson(res);
}

export async function runEngineeringReview(buildDir, { force = false, timeoutS = 180 } = {}) {
  const res = await fetch(`${API_BASE}/v1/build-files/engineering-review/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      build_dir: buildDir,
      force,
      timeout_s: timeoutS,
    }),
  });
  return parseJson(res);
}
